### PR TITLE
BUG: Fix fails of some tests and bad behaviour for x=zeta in levy_stable

### PR DESCRIPTION
**Pull request by nokados**
#### Reference issue
#17493 

#### What does this implement/fix?
It actually contains 2 contributions:
- fix of failures of levy_stable tests caused by changes of global parameters
- fix of incorrect pdf values in x = 0 (x near to zeta)

#### Additional information
I think the problems with incorrect pdf values in x = 0 may vary for different environments. They caused by differences in zeta calculated in Python and zeta calculated in C (Nolan), whereas the code assumes them equal. So, for different OS, Python, processor architectures, the proposed tests may be insufficient. I tested on x86_64, Linux 5.4.0-132-generic, Ubuntu 20.04.5 LTS, Python 3.10.10.

### CODE REVIEW

**Comment by melissawm** [comment]
Thanks for the PR @nokados ! It doesn't look like the CI failures are related to your changes.

**Comment by nokados** [comment]
So, do we need some additional work here or PR is ready to be merged?

**Comment by mdhaber** [comment]
The PR needs to be reviewed by a maintainer before it can be merged. That may take some time. @steppi has experience with `levy_stable` so I'll ping him to see if he's interested.

**Comment by steppi** [comment]
> The PR needs to be reviewed by a maintainer before it can be merged. That may take some time. @steppi has experience with `levy_stable` so I'll ping him to see if he's interested.

Pretty busy right now, but I should have some spare time in a couple weeks.

**Comment by nokados** [comment]
This review should not be difficult, it doesn't require math or complicated algorithms or structures... Just want to remember :)

**Comment by nokados** [comment]
> Thanks for the reminder @nokados. I haven’t had a spare moment but hopefully should have some time over the weekend. 

Any progress here? If you are still busy, maybe, do you want to delegate the review to someone else?

**Comment by steppi** [comment]
> > Thanks for the reminder @nokados. I haven’t had a spare moment but hopefully should have some time over the weekend.
> 
> Any progress here? If you are still busy, maybe, do you want to delegate the review to someone else?

Thanks for reminding me. I'll look at it today. As of this week I'm no longer swamped with things to do. Sorry for the long delay and thanks for your patience.

**Review by steppi — CHANGES_REQUESTED** [review]
I've checked carefully through the code and it looks like the changes to `pdf` and `cdf` calculation amount only to adding some no-ops. The new function `_nolan_round_x_near_zero` is only called after `_nolan_round_difficult_input` has already been called. If you had tests failing locally, I suspect it may have only been because the `levy_stable` class variables were not being reset between test runs.

I suggest to remove all of the changes here, except to keep the fixture which resets the class variables. The PR name should then be changed to something like
`TST: Reset levy_stable params between tests`

File: scipy/stats/_levy_stable/__init__.py
Comment (nokados) [reply scipy/stats/_levy_stable/__init__.py:233]: Yes, the rounding has already occurred, however the value of zeta changes later (line 228). So, the condition at line 233 `if x0 == zeta:` becomes false. This leads to incorrect answers that surprised me when I saw the negative values of PDF at the points where I expected to see the peak. So we really need to do rounding here. Maybe, we could eliminate the first rounding at `_pdf_single_value_piecewise_Z0`, but doing this caused fails of some tests for me.

File: scipy/stats/tests/test_distributions.py
Comment (nokados) [reply scipy/stats/tests/test_distributions.py:4248]: I relied on an example from [SO](https://stackoverflow.com/questions/22627659/run-code-before-and-after-each-test-in-py-test). I'll try without it

File: scipy/stats/tests/test_distributions.py
Comment (nokados) [reply scipy/stats/tests/test_distributions.py:4975]: Yeah, it really tests cases only for `x0 == zeta`. This case fails too, but I'll also add tests for `x0 != zeta`

File: scipy/stats/_levy_stable/__init__.py
Comment (steppi) [reply scipy/stats/_levy_stable/__init__.py:233]: Oh, right. sorry about that. I'll need to take a closer look. It's strange that your getting these failures on such a common platform. I'd like to be able to reproduce this locally.

```I tested on x86_64, Linux 5.4.0-132-generic, Ubuntu 20.04.5 LTS, Python 3.10.10.```

Comment (steppi) [reply scipy/stats/_levy_stable/__init__.py:233]: In Python zeta is calculated like this

```python
zeta = -beta * np.tan(np.pi * alpha / 2.0)
````

and in C like this

```C
 sp->zeta = -beta * tan(M_PI_2 * alpha);
```

Since the formula's aren't quite the same, C it's not surprising that these values could be different on some platforms. Your right, we shouldn't be assuming that the value computed in C will be the same as that computed in Python

Comment (steppi) [reply scipy/stats/_levy_stable/__init__.py:233]: Your solution works, I wonder if there's a clearer way to do it though. If we go with yours, we'll need a comment explaining what's happening

Comment (nokados) [reply scipy/stats/_levy_stable/__init__.py:233]: As an alternative I see
1) Using the old calculated zeta
2) Comparing "x close to zeta" instead of "x == zeta"

Unfortunately, I am not experienced enough to predict do these solutions lead to unexpected behaviour in other places of the algorithm or not.

Comment (nokados) [reply scipy/stats/_levy_stable/__init__.py:233]: > What C compiler are you using?

Cython 0.29.33
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

To reproduce the error, I advise you to reduce `stats.levy_stable.quad_eps`, for example
```python
stats.levy_stable.quad_eps = 1.2e-11
```
Also, you can use this snippet to search for bad parameters:
```python
while True:
    alpha = np.random.uniform(0, 2, size=2000)
    beta = np.random.uniform(-1, 1, size=2000)
    logpdf = levy_stable.logpdf(0, alpha=alpha, beta=beta)
    if pd.isna(logpdf).any():
        bad = pd.DataFrame(
            dict(
                alpha=alpha[pd.isna(logpdf)],
                beta=beta[pd.isna(logpdf)]
            )
        )
        print('Found! ', bad)
        break
```

**Review by steppi — CHANGES_REQUESTED** [review]
OK, just a few minor suggestions but overall I think this looks good.

**Comment by steppi** [comment]
@nokados Can you also merge main into your branch? There have been some CI changes.

File: scipy/stats/tests/test_distributions.py
Comment (steppi) [reply scipy/stats/tests/test_distributions.py:4965]: Actually, it looks like we've changed our line length standards and the new max length is 88 so this is allowed https://github.com/scipy/scipy/issues/18347#issuecomment-1518830905

Was there a reason you wrote out 20 digits after the decimal place? I still think just including 16 digits in these float literals is better, but you can leave it as it is if you like.

Comment (steppi) [reply scipy/stats/tests/test_distributions.py:4965]: Sorry, it should be 17 digits. [Machine Epsilon](https://en.wikipedia.org/wiki/Machine_epsilon) is an upper bound on floating point approximation error, but for some numbers you can get up to 17 digits of accuracy.

Comment (nokados) [reply scipy/stats/tests/test_distributions.py:4965]: Sorry for the delay. The suggestion with the shorter floats works well. I'm adding it to the commit

File: scipy/stats/tests/test_distributions.py
Comment (nokados) [reply scipy/stats/tests/test_distributions.py:4970]: Actually, I generated the expected values with the fixed version of scipy code. Could I assume that it was really fixed correctly and use this values? Should I double-check the values using Nolan's C program or some else?

Comment (steppi) [reply scipy/stats/tests/test_distributions.py:5218]: Yes, out of principle you should compare to something external.

Comment (nokados) [reply scipy/stats/tests/test_distributions.py:5218]: OK. I don't want to use Nolan's program, because a free version of his program is Windows-only and closed-source. So I tested in an online version of [MATLAB](https://www.mathworks.com/help/stats/stable-distribution.html) (ironically, isn't?) using such code:
```matlab
format long

alpha = 1.5748002527689913;
beta = -0.25200194914153684;
expected = 0.28013657621866517;

pd = makedist('Stable','alpha',alpha,'beta',beta,'gam',1,'delta',0);
% we need to adjust x. It is the same as x = 0 In scipy.
x = -beta * tan(pi * alpha / 2)
pd.pdf(x)
assert (expected - pd.pdf(x) <= eps) % eps is constant - machine epsilon
```

All tests have passed successfully. Is it good enough? Do I still need to add the comment?

**Comment by nokados** [comment]
There are CI failures even after merge with main. Do we have to do something about it?

**Comment by j-bowhay** [comment]
The doc build was broken yesterday so you were probably just unfortunate when you merged main. Merging again should fix the issue

File: scipy/stats/tests/test_distributions.py
Comment (steppi) [reply scipy/stats/tests/test_distributions.py:5218]: MATLAB’s as a reference implementation is fine. I think you should change the expected values in the relevant SciPy test to the values produced by MATLAB, and leave a comment stating where the values came from.

**Comment by steppi** [comment]
@nokados this is ready to go, except the test should be changed to use the expected values from MATLAB, and a comment should be added stating this is where they came from. I know that the expected values closely match MATLAB, and it might seem like a formality, but I don't think it's OK to test the function against itself. If you don't have time at the moment I can make the changes if you like.

**Comment by nokados** [comment]
> @nokados this is ready to go, except the test should be changed to use the expected values from MATLAB, and a comment should be added stating this is where they came from. I know that the current expected values closely match MATLAB, and it might seem like a formality, but I don't think it's OK to test the function against itself. If you don't have time at the moment I can make the changes if you like.

@steppi I understand, you're right, it should be changed. Yeah, sorry, I don't have any spare time right now. Maybe I'll be free in 2 weeks. If you help me with the changes, I'll be glad and appreciate it.

**Comment by steppi** [comment]
> > @nokados this is ready to go, except the test should be changed to use the expected values from MATLAB, and a comment should be added stating this is where they came from. I know that the current expected values closely match MATLAB, and it might seem like a formality, but I don't think it's OK to test the function against itself. If you don't have time at the moment I can make the changes if you like.
> 
> @steppi I understand, you're right, it should be changed. Yeah, sorry, I don't have any spare time right now. Maybe I'll be free in 2 weeks. If you help me with the changes, I'll be glad and appreciate it.

No worries, I can do it. I appreciate your help finding and fixing this one.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5212]:
```suggestion
            (1.7720732804618808, 0.5059373136902996, 0.278932636798268),
            (1.9217001522410235, -0.8779442746685926, 0.281054757202316),
            (1.5654806051633634, -0.4016220341911392, 0.271282133194204),
            (1.7420803447784388, -0.38180029468259247, 0.280202199244247),
            (1.5748002527689913, -0.25200194914153684, 0.280136576218665),
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5207]:
```suggestion
        'alpha,beta,expected',
        # Reference values computed with Matlab, e.g.
        # format long
        # alpha = 1.5748002527689913;
        # beta = -0.25200194914153684;
        # pd = makedist('Stable','alpha',alpha,'beta',beta,'gam',1,'delta',0);
        # % we need to adjust x. It is the same as x = 0 in scipy.
        # x = -beta * tan(pi * alpha / 2)
        # pd.pdf(x)
        [
```

**Comment by mdhaber** [comment]
I went ahead and implemented https://github.com/scipy/scipy/pull/18133#issuecomment-1564932874.

**Comment by steppi** [comment]
> I went ahead and implemented [#18133 (comment)](https://github.com/scipy/scipy/pull/18133#issuecomment-1564932874). (Just pushing buttons; no review here!)
> 
> Is this intended to close [gh-17493](https://github.com/scipy/scipy/issues/17493), or was that just intended to be mentioned?

Thanks. Yes it does fix gh-17493. The issue there was that class variables for the distribution were not being reset between test runs. It’s one of the two fixes in this PR.

**Comment by steppi** [comment]
> I went ahead and implemented [#18133 (comment)](https://github.com/scipy/scipy/pull/18133#issuecomment-1564932874). (Just pushing buttons; no review here!)
> 
> Is this intended to close [gh-17493](https://github.com/scipy/scipy/issues/17493), or was that just intended to be mentioned?

Oh, also, I tried to get the expected values from the R `stabledist` package and they were different for the, `x != 0` cases, and those tests actually failed. Did you use matlab here, can you check those cases too? I wasn’t sure if the problem was on our side or R’s.

**Comment by mdhaber** [comment]
Where are the `x != 0` cases? In `test_distributions::TestLevyStable`?

**Comment by steppi** [comment]
> Where are the `x != 0` cases? In `test_distributions::TestLevyStable`?

There are actually three asserts in `test_x_near_zeta`, two for $x = \pm 0.0001$.  I can push my version of the test, I should have time to do it in around 20 minutes.

**Comment by mdhaber** [comment]
Sure; I won't get to it by then.

**Comment by steppi** [comment]
> Can you edit the script in [#18133 (comment)](https://github.com/scipy/scipy/pull/18133#discussion_r1196970769) with one of those test cases so that I know that I'm putting the numbers into the right place? If so, I can do the rest.

Sure. It amounts to converting from Nolan's `S1` parametrization to his `S0` parametrization

```matlab

format long

alpha = 1.5748002527689913;
beta = -0.25200194914153684;
expected = 0.28013657621866517;

pd = makedist('Stable','alpha',alpha,'beta',beta,'gam',1,'delta',0);
% we need to adjust x since matlab only handles S0 parametrization
x = 0.0001 - beta * tan(pi * alpha / 2)
pd.pdf(x)
assert (expected - pd.pdf(x) <= eps) % eps is constant - machine epsilon
```

**Comment by mdhaber** [comment]
And for the cases with `-1e-4`, just replace `1e-4` with`-1e-4`? (`- beta * tan(pi * alpha / 2)` is a constant offset?)

**Comment by steppi** [comment]
> 

Yes. That looks right.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5222]:
```suggestion
            (0, 1.7720732804618808, 0.5059373136902996, 0.278932636798268),
            (0, 1.9217001522410235, -0.8779442746685926, 0.281054757202316),
            (0, 1.5654806051633634, -0.4016220341911392, 0.271282133194204),
            (0, 1.7420803447784388, -0.38180029468259247, 0.280202199244247),
            (0, 1.5748002527689913, -0.25200194914153684, 0.280136576218665),
            (1e-4, 1.7720732804618808, 0.5059373136902996, 0.278929165340670),
            (1e-4, 1.9217001522410235, -0.8779442746685926, 0.281056564327953),
            (1e-4, 1.5654806051633634, -0.4016220341911392, 0.271252432161167),
            (1e-4, 1.7420803447784388, -0.38180029468259247, 0.280205311264134),
            (1e-4, 1.5748002527689913, -0.25200194914153684, 0.280140965235426),
            (-1e-4, 1.7720732804618808, 0.5059373136902996, 0.278936106741754),
            (-1e-4, 1.9217001522410235, -0.8779442746685926, 0.281052948629429),
            (-1e-4, 1.5654806051633634, -0.4016220341911392, 0.271275394392385),
            (-1e-4, 1.7420803447784388, -0.38180029468259247, 0.280199085645099),
            (-1e-4, 1.5748002527689913, -0.25200194914153684, 0.280132185432842),
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5207]:
```suggestion
        'x,alpha,beta,expected',
        # Reference values from Matlab
        # format long
        # alphas = [1.7720732804618808, 1.9217001522410235, 1.5654806051633634,
        #           1.7420803447784388, 1.5748002527689913];
        # betas = [0.5059373136902996, -0.8779442746685926, -0.4016220341911392,
        #          -0.38180029468259247, -0.25200194914153684];
        # x0s = [0, 1e-4, -1e-4];
        # for x0 = x0s
        #     disp("x0 = " + x0)
        #     for ii = 1:5
        #         alpha = alphas(ii);
        #         beta = betas(ii);
        #         pd = makedist('Stable','alpha',alpha,'beta',beta,'gam',1,'delta',0);
        #         % we need to adjust x. It is the same as x = 0 In scipy.
        #         x = x0 - beta * tan(pi * alpha / 2);
        #         disp(pd.pdf(x))
        #     end
        # end
        [
```

**Comment by steppi** [comment]
Thanks. Also @nokados had to lower `quad_eps` to get the tests to fail, which will lower the accuracy of our results. I'm going to try again with the default values, but it will take a moment to build scipy.

File: scipy/stats/tests/test_distributions.py
Comment (steppi) [inline scipy/stats/tests/test_distributions.py:5263]:
```suggestion
    @pytest.mark.parametrize(
        'x,alpha,beta,expected',
        # Reference values from Matlab
        # format long
        # alphas = [1.7720732804618808, 1.9217001522410235, 1.5654806051633634,
        #           1.7420803447784388, 1.5748002527689913];
        # betas = [0.5059373136902996, -0.8779442746685926, -0.4016220341911392,
        #          -0.38180029468259247, -0.25200194914153684];
        # x0s = [0, 1e-4, -1e-4];
        # for x0 = x0s
        #     disp("x0 = " + x0)
        #     for ii = 1:5
        #         alpha = alphas(ii);
        #         beta = betas(ii);
        #         pd = makedist('Stable','alpha',alpha,'beta',beta,'gam',1,'delta',0);
        #         % we need to adjust x. It is the same as x = 0 In scipy.
        #         x = x0 - beta * tan(pi * alpha / 2);
        #         disp(pd.pdf(x))
        #     end
        # end
        [
            (0, 1.7720732804618808, 0.5059373136902996, 0.278932636798268),
            (0, 1.9217001522410235, -0.8779442746685926, 0.281054757202316),
            (0, 1.5654806051633634, -0.4016220341911392, 0.271282133194204),
            (0, 1.7420803447784388, -0.38180029468259247, 0.280202199244247),
            (0, 1.5748002527689913, -0.25200194914153684, 0.280136576218665),
        ]
    )
    def test_x_equal_zeta(
            self, x, alpha, beta, expected
    ):
        """Test pdf for x equal to zeta.

        With S1 parametrization: x0 = x + zeta if alpha != 1 So, for x = 0, x0
        will be close to zeta.

        When case "x equal zeta" is not handled properly and quad_eps is not
        low enough: - pdf may be less than 0 - logpdf is nan

        The points from the parametrize block are found randomly so that PDF is
        less than 0.

        Reference values taken from MATLAB
        https://www.mathworks.com/help/stats/stable-distribution.html
        """
        stats.levy_stable.quad_eps = 1.2e-11

        assert_almost_equal(
            stats.levy_stable.pdf(x, alpha=alpha, beta=beta),
            expected,
        )

    @pytest.mark.xfail
    @pytest.mark.parametrize(
        # See comment for test_x_equal_zeta for script for reference values
        'x,alpha,beta,expected',
        [
            (1e-4, 1.7720732804618808, 0.5059373136902996, 0.278929165340670),
            (1e-4, 1.9217001522410235, -0.8779442746685926, 0.281056564327953),
            (1e-4, 1.5654806051633634, -0.4016220341911392, 0.271252432161167),
            (1e-4, 1.7420803447784388, -0.38180029468259247, 0.280205311264134),
            (1e-4, 1.5748002527689913, -0.25200194914153684, 0.280140965235426),
            (-1e-4, 1.7720732804618808, 0.5059373136902996, 0.278936106741754),
            (-1e-4, 1.9217001522410235, -0.8779442746685926, 0.281052948629429),
            (-1e-4, 1.5654806051633634, -0.4016220341911392, 0.271275394392385),
            (-1e-4, 1.7420803447784388, -0.38180029468259247, 0.280199085645099),
            (-1e-4, 1.5748002527689913, -0.25200194914153684, 0.280132185432842),
        ]
    )
    def test_x_near_zeta(
            self, x, alpha, beta, expected
    ):
        """Test pdf for x near zeta.

        With S1 parametrization: x0 = x + zeta if alpha != 1 So, for x = 0, x0
        will be close to zeta.

        When case "x near zeta" is not handled properly and quad_eps is not
        low enough: - pdf may be less than 0 - logpdf is nan

        The points from the parametrize block are found randomly so that PDF is
        less than 0.

        Reference values taken from MATLAB
        https://www.mathworks.com/help/stats/stable-distribution.html
        """
        stats.levy_stable.quad_eps = 1.2e-11

        assert_almost_equal(
            stats.levy_stable.pdf(x, alpha=alpha, beta=beta),
            expected,
        )

```

**Review by steppi — APPROVED** [review]
OK, I think this is in good shape now.

To summarize, there are two changes here:

1. Fix the tests by resetting the class variables for `levy_stable` between test runs. This was causing non-deterministic failures observed in gh-17493.
2. Fix an issue involving `levy_stable` rounding `x0` to `zeta` when `x0` is sufficiently near `zeta`. After the rounding, it is assumed that `x0` will always equal `zeta`, but at one point in the code, `zeta` is recomputed in a `C` extension, and on some platforms, the value computed in `C` can be a one or two ulps off from the value computed in `Python`. The mistaken assumption was causing the density to evaluate to negative in some cases on some platforms. @nokados solution is to round `x0` to `zeta` again if needed after `zeta` is recomputed, which I think is acceptable.

**Comment by steppi** [comment]
Thanks @nokados. Congrats on getting your first PR merged!

**Pull request merged**
